module/zfs/dsl_bookmark.c

```diff
@@ -413,6 +427,148 @@ dsl_bookmark_create(nvlist_t *bmarks, nvlist_t *errors)
 	    fnvlist_num_pairs(bmarks), ZFS_SPACE_CHECK_NORMAL));
 }
 
+typedef struct dsl_bookmark_clone_arg {
+	nvlist_t *dbcc_bmarks;
+	nvlist_t *dbcc_errors;
+} dsl_bookmark_clone_arg_t;
+
+static int
+dsl_bookmark_clone_check_impl(nvpair_t *pair, dmu_tx_t *tx)
+{
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	int error;
+	zfs_bookmark_phys_t bmark_phys = { 0 };
+
+	/* Verify that the new bookmark does not already exist */
+	error = dsl_bookmark_lookup(dp, nvpair_name(pair), NULL, &bmark_phys);
+	switch (error) {
+	case 0:
+		return (SET_ERROR(EEXIST));
+	case ESRCH:
+		/* happy path: new bmark doesn't exist, proceed after switch */
+		break;
+	default:
+		return (error);
+	}
+
+	/* Verify that the target bookmark exists */
+	error = dsl_bookmark_lookup(dp, fnvpair_value_string(pair), NULL, &bmark_phys);
+	if (error != 0) {
+		return (error);
+	}
+
+	return (0);
+}
+
+static int
+dsl_bookmark_clone_check(void *arg, dmu_tx_t *tx)
+{
+	dprintf("dsl_bookmark_clone_check began\n");
+	dsl_bookmark_clone_arg_t *dbcc = arg;
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	int rv = 0;
+
+	if (!spa_feature_is_enabled(dp->dp_spa, SPA_FEATURE_BOOKMARKS))
+		return (SET_ERROR(ENOTSUP));
+
+	for (nvpair_t *pair = nvlist_next_nvpair(dbcc->dbcc_bmarks, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(dbcc->dbcc_bmarks, pair)) {
+		int error;
+
+		/*
+		 * We assume that dbcc_bmarks has unique names (keys)
+		 * This is enforced by zfs_ioc_bookmark_clone
+		 */
+
+		error = dsl_bookmark_clone_check_impl(pair,  tx);
+		if (error != 0) {
+			fnvlist_add_int32(dbcc->dbcc_errors, nvpair_name(pair), error);
+			rv = error;
+		}
+	}
+
+	return (rv);
+}
+
+static void
+dsl_bookmark_clone_sync_impl(const char *new_name, const char *target_name,
+    dmu_tx_t *tx, uint64_t num_redact_snaps, uint64_t *redact_snaps, void *tag,
+    redaction_list_t **redaction_list)
+{
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	dsl_dataset_t *bmark_fs_target, *bmark_fs_new;
+	char *target_shortname, *new_shortname;
+	zfs_bookmark_phys_t target_phys;
+
+	// TODO need hold pool? dsl_get_bookmark_props makes it seem like we should...?
+
+	VERIFY0(dsl_bookmark_hold_ds(dp, target_name, &bmark_fs_target, FTAG,
+	    &target_shortname));
+	VERIFY0(dsl_bookmark_hold_ds(dp, new_name, &bmark_fs_new, FTAG,
+	    &new_shortname));
+
+	// TODO assert bmark_fs_target's filesystem is the same as bmark_fs_new's filesystem
+
+	/* copy target phys to new phys */
+	VERIFY0(dsl_bookmark_lookup_impl(bmark_fs_target, target_shortname, &target_phys));
+	dsl_bookmark_node_t *new_dbn = dsl_bookmark_node_alloc(new_shortname);
+	dsl_bookmark_copy_phys(&new_dbn->dbn_phys, &target_phys);
+
+	// TODO need to take care of take care of redaction bookmarks?
```

Comment:
Do we really want to copy the redaction object?

AFAICT it's immutable once it is created through `zfs redact`, and the redaction bookmark doesn't exist until after `zfs redact` completes,  which means we cannot come into a situation where we would try to copy an incomplete redaction bookmark.

=> can't we just have a reference count for this? (obviously, for O(1) bookmark deletion, this would need to be tracked on disk

Reply:
That should be fine as well.